# check 2012 chemoproteomic pos/protein sequence against 2018 human proteome reference

Pmap project 

11.10.19

mfpfox

## input files: 
**from 2012**
- UKBID_2012reference_4084.csv
- cravattlab_11_2012.csv (needs to be filtered for IDs)

**from 2018**
- UKB_CCDS_2018_06_fasta.csv

## outputfiles: 
- AllReact2018_5856.csv
- ALLReact2018_grouped_entry_threshold_2751.csv
- CLMH_description_table_2018.csv
- Cys2018_ever_labeled_6234.csv
- CYS2018_everlabeled_gene_valueCounts_2915.csv
- Cys2018_pantarget_5962.csv
- Cys2018_reactive_1431.csv
- dictOfPositions_allCysLys_3964.csv
- KLMH_description_table_2018.csv
- Lys2018_ever_labeled_9046.csv
- LYS2018_everlabeled_gene_valueCounts_2624.csv
- Lys2018_pantarget_8115.csv
- Lys2018_reactive_4425.csv
- UKB2018_reference_dict_mismap_3964.csv
- UKB2018_reference_proteome_mappedCKpositions_3964.csv

In [64]:
import os
import sys
import pandas as pd
import numpy as np
from ast import literal_eval 

In [2]:
sys.path.append("/Users/mariapalafox/Desktop/Toolbox")
from all_funx import *

In [3]:
from IPython.display import display
pd.set_option('display.max_columns', None)
pd.options.display.max_seq_items = 2000

from IPython.core.display import display, HTML
display(HTML("<style>.container {width:90% !important;}</style>"))

In [4]:
os.chdir("/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CYS_LYS_2012_to2018UKBpositions/")
print(os.listdir())

['Lys2012_reactive_4501.csv', 'CYS2012_everlabeled_gene_valueCounts_3006.csv', 'UKBID_2012reference_4084.csv', 'CCDS2018_all_Cpos.csv', 'AllReact2012_5959.csv', '.DS_Store', 'LYS2012_everlabeled_gene_valueCounts_2683.csv', 'ALLReact2012_grouped_entry_threshold_2807.csv', 'CravattLAB', 'MAPPED_CK_v1', 'Lys2012_pantarget_8257.csv', 'NATURE2012_UKBccds2018_files', 'POS_NOT_FOUND_DROP_THESE_IDS.csv', 'CYSLYS2012_everlabeled_gene_valueCounts_4084.csv', 'CCDS2018_all_Kpos.csv', 'Cys2012_ever_labeled_6404.csv', '2012_01_UKB_Swiss_release', 'CLMH_description_table_2012.csv', 'Reprocesssing2012_Nature_allCYSLYSdata.xlsx', 'UKB_CCDS_2018_06_fasta.csv', 'all2012_datasets_not_filtered', 'KLMH_description_table_2012.csv', 'Cys2012_pantarget_6125.csv', 'Cys2012_reactive_1458.csv', '2013_01_UKB_Swiss_release', 'Lys2012_ever_labeled_9213.csv', 'cravattlab_11_2012.csv', 'UniqueIDs_og_data_2012_11_uniprot_sprot_4119.csv']


# [0] filter fasta sequences from 2012 for IDs contained in 4084 list 
- (notes on 2012 datasets in part A markdown, counts in Mapping_library)

In [5]:
ids = pd.read_csv("UKBID_2012reference_4084.csv")
fasta12 = pd.read_csv("cravattlab_11_2012.csv")
print(ids.shape)
print(fasta12.shape)
print()
fasta18 = pd.read_csv("UKB_CCDS_2018_06_fasta.csv")
print(fasta18.shape)

(4084, 1)
(19981, 4)

(18432, 5)


In [10]:
# same as above but also drops False value rows
def addcolumnconditionalDrop(mapList, df, dfcol, newcol): 
    mendel = []
    for g in df[dfcol]:
        if g in mapList: 
            mendel.append("True")
        else:
            mendel.append("False")
    df.loc[:, newcol] = mendel
    df.drop(df[df[newcol] == "False"].index, inplace = True)
    df.reset_index(inplace=True, drop=True)
    print(df.shape)
    return df

In [11]:
fasta12.head(3)

ID    entryName  Length  \
0  Q8NGR1  O13A1_HUMAN     328   
1  Q8N6Q8  MET25_HUMAN     603   
2  Q9UBK2  PRGC1_HUMAN     798   

                                         proSequence  
0  MKLWMESHLIVPETRPSPRMMSNQTLVTEFILQGFSEHPEYRVFLF...  
1  MAASCPLPVTPDLPTLRAKLQGLLQFLRDALSISNAHTVDFYTESV...  
2  MAWDMCNQDSESVWSDIECAALVGEDQPLCPDLPELDLSELDVNDL...

In [9]:
dropUnnamed(fasta18)
fasta18.head(3)

ID    entryName  Length  \
0  O95139  NDUB6_HUMAN     128   
1  O75438  NDUB1_HUMAN      58   
2  Q8N4C6    NIN_HUMAN    2090   

                                         proSequence  
0  MTGYTPDEKLRLQQLRELRRRWLKDQELSPREPVLPPQKMGPMEKF...  
1  MVNLLQIVRDHWVHVLVPMGFVIGCYLDRKSDERLTAFRNKSMLFK...  
2  MDEVEQDQHEARLKELFDSFDTTGTGSLGQEELTDLCHMLSLEEVA...

In [12]:
idlist = list(ids.ID)
seq2012 = addcolumnconditionalDrop(idlist, fasta12, 'ID','in4084')
seq2018 = addcolumnconditionalDrop(idlist, fasta18, 'ID', 'in4084')

(4084, 5)
(3976, 5)


In [14]:
4084-3976

108

In [15]:
id12 = set(seq2012.ID)
id18 = set(seq2018.ID)

diff = list(id12 - id18)
print(len(diff))
print(diff)

108
['E7EVH7', 'E7EQZ4', 'O00422', 'Q9NVE7', 'Q53H96', 'P05534', 'P16455', 'O60361', 'P49189', 'P42858', 'Q9H7T3', 'P0CB43', 'Q6P4F2', 'Q5JZ73', 'P49411', 'Q9H9E3', 'A0JLT2', 'H3BQZ7', 'Q9Y4D8', 'Q6NZY4', 'H3BN98', 'P56556', 'Q7Z3J2', 'O75592', 'Q7Z460', 'J3QR44', 'P0CG05', 'P29034', 'P16435', 'F8VZW7', 'E2QRD5', 'Q99836', 'Q9NX70', 'Q8IY67', 'Q16181', 'B0V043', 'H0Y2S0', 'E7EVK2', 'P62158', 'Q9BWM7', 'P57053', 'Q6ZNA5', 'Q9H3P2', 'Q8NAG6', 'O75122', 'P09525', 'Q96I15', 'Q6DKI1', 'P62333', 'I3L499', 'Q9Y618', 'P30460', 'J3KR12', 'Q9Y5J9', 'C9J4G0', 'P01871', 'P01892', 'P0DI82', 'Q9H6E5', 'P08107', 'Q5VYK3', 'Q86YG4', 'E7ETI0', 'P30085', 'Q9Y6A9', 'H3BQ06', 'Q9Y5A7', 'P61758', 'Q8IXH7', 'P17405', 'P0CW22', 'Q8WYJ6', 'Q96EY9', 'Q8N5W9', 'P49454', 'O43504', 'Q5VTU8', 'Q86V81', 'A6NE09', 'A2A2Z9', 'Q9NYZ3', 'Q8NDA2', 'Q9H1Z4', 'D6RAD4', 'P62861', 'I3L420', 'P24928', 'P07988', 'Q8N9T8', 'H0YGG7', 'P30042', 'F5H5P2', 'Q9H3K6', 'P26440', 'Q8WX92', 'Q9Y4K4', 'P55145', 'H3BVE0', 'H7BZ11', 'A8MV

## REASONS why 108 IDs not in 2018 but in 2012 fasta: 
1. ID not apart of CCDS set, which is filtered used to make 2018 reference proteome (uncharacterized pro example)
2. ID not apart of Human Proteome set of IDs in UKB
3. ID no longer in UKB
4. ID merged with another to form new ID
5. ID was demerged to form two new IDs

# [1] identity check between 2012 sequences and 2018 sequences

In [18]:
seq2012.head(4)

ID    entryName  Length  \
0  P35998   PRS7_HUMAN     433   
1  Q96L92  SNX27_HUMAN     541   
2  Q2WGJ9  FR1L6_HUMAN    1857   
3  Q99590  SCAFB_HUMAN    1463   

                                         proSequence in4084  
0  MPDYLGADQRKTKEDEKDDKPIRALDEGDIALLKTYGQSTYSRQIK...   True  
1  MADEDGEGIHPSAPHRNGGGGGGGGSGLHCAGNGGGGGGGPRVVRI...   True  
2  MFGLKVKKKRNKAEKGLILANKAAKDSQGDTEALQEEPSHQEGPRG...   True  
3  MKKKTVCTLNMGDKKYEDMEGEENGDNTISTGLLYSEADRCPICLN...   True

In [19]:
seq2018.head(4)

ID    entryName  Length  \
0  Q9HAS0   NJMU_HUMAN     396   
1  Q86X76   NIT1_HUMAN     327   
2  Q9NQR4   NIT2_HUMAN     276   
3  Q9Y314  NOSIP_HUMAN     301   

                                         proSequence in4084  
0  MLPSLQESMDGDEKELESSEEGGSAEERRLEPPSSSHYCLYSYRGS...   True  
1  MLGFITRPPHRFLSLLCPGLRIPQLSVLCAQPRPRAMAISSSSCEL...   True  
2  MTSFRLALIQLQISSIKSDNVTRACSFIREAATQGAKIVSLPECFN...   True  
3  MTRHGKNCTAGAVYTYHEKKKDTAASGYGTQNIRLSRDAVKDFDCC...   True

In [37]:
# assumes ID and proSequence column
# compares the sequence in df and dict value, adds column based on comparison results
# True is they are identical, False is they are different, no info on HOW diff they are
def identicalSequenceCheck(dfref, dfalt):
    ref_dic = dict(zip(dfref.ID, dfref.proSequence))
    ref_len = dict(zip(dfref.ID, dfref.Length))
    newcol = []
    len2012 = []
    for index, row in dfalt.iterrows():
        pep = row['proSequence']
        ukb_id = row['ID']
        # retrieve uniprot fasta seq using reference dictionary
        mypep = ref_dic[ukb_id]
        len12 = ref_len[ukb_id]
        str(mypep)
        if mypep == pep:
            newcol.append('True')
            len2012.append(len12)
        if mypep != pep: 
            newcol.append('False')
            len2012.append(len12)
    # add identity score list as new column
    dfalt.loc[:,'identical_seq'] = newcol
    dfalt.loc[:,'length_2012_fasta'] = len2012
    dfalt.loc[:,'len12_minus_len18'] = dfalt['length_2012_fasta'] - dfalt['Length']
    print(dfalt.shape)
    checkColumnValues(dfalt, "identical_seq")
    return dfalt

In [39]:
checked2018 = identicalSequenceCheck(seq2012, seq2018)

(3976, 8)
  identical_seq  Count
0          True   3931
1         False     45


In [50]:
checked2018.to_csv("UKBref_2018_T3931_F45.csv", index=False)

In [40]:
false45 = checked2018[checked2018['identical_seq'] == 'False']

In [46]:
# assumes ID and proSequence column
# $$$
def sequenceDifference(dfref, dfalt):
    ref_dic = dict(zip(dfref.ID, dfref.proSequence))
    diffcol = []
    for index, row in dfalt.iterrows():
        pep = row['proSequence']
        ukb_id = row['ID']
        # retrieve uniprot fasta seq using reference dictionary
        mypep = ref_dic[ukb_id]
        str(mypep)
        if mypep == pep:
            diffcol.append('nodiff')
        if mypep != pep: 
            output_list = [li for li in difflib.ndiff(mypep, pep) if li[0] != ' ']
            out = ', '.join(output_list)
            diffcol.append(out)
    # add identity score list as new column
    dfalt.loc[:,'difference_12vs18'] = diffcol
    print(dfalt.shape)
    return dfalt

In [47]:
diff2018 = sequenceDifference(seq2012, false45)

(45, 9)


/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:362: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[key] = _infer_fill_value(value)
/anaconda3/lib/python3.7/site-packages/pandas/core/indexing.py:543: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self.obj[item] = s


In [48]:
diff2018

ID    entryName  Length  \
24    P14598   NCF1_HUMAN     390   
261   Q96CV9   OPTN_HUMAN     577   
328   Q86YV5  PRAG1_HUMAN    1406   
333   Q9H307  PININ_HUMAN     717   
343   Q9NPH0   PPA6_HUMAN     428   
414   Q15283  RASA2_HUMAN     850   
423   Q9UKL0  RCOR1_HUMAN     485   
439   P20338  RAB4A_HUMAN     218   
551   Q6WKZ4  RFIP1_HUMAN    1283   
563   Q16822  PCKGM_HUMAN     640   
701   Q9NXR1   NDE1_HUMAN     335   
748   P20929   NEBU_HUMAN    6669   
796   Q5JRX3   PREP_HUMAN    1037   
865   Q99996  AKAP9_HUMAN    3907   
947   Q9Y2K2   SIK3_HUMAN    1321   
962   P04179   SODM_HUMAN     222   
1006  Q9NUL3  STAU2_HUMAN     570   
1251  P82921   RT21_HUMAN      87   
1345  O75044  SRGP2_HUMAN    1071   
1419  Q9P2E9  RRBP1_HUMAN    1410   
1456  Q96BW9  TAM41_HUMAN     452   
1673  P06753   TPM3_HUMAN     285   
1727  Q86UA6    RIP_HUMAN     219   
1844  P80303  NUCB2_HUMAN     420   
1954  Q8WUH1   CHUR_HUMAN     139   
1993  Q96B23  CR025_HUMAN     404   
2074  Q5T0N5  FBP1L_HUMAN     605   
2223  Q13315    ATM_HUMAN    3056   
2234  Q66PJ3  AR6P4_HUMAN     421   
2395  Q99873   ANM1_HUMAN     371   
2474  P13929   ENOB_HUMAN     434   
2504  Q6P3S1  DEN1B_HUMAN     775   
2545  Q5VZ89  DEN4C_HUMAN    1909   
2626  Q9Y680  FKBP7_HUMAN     222   
2637  Q9BXW9  FACD2_HUMAN    1451   
2930  Q14643  ITPR1_HUMAN    2758   
2948  Q8IV33  K0825_HUMAN    1275   
3001  Q13555  KCC2G_HUMAN     558   
3002  Q9NQT8  KI13B_HUMAN    1826   
3053  Q03252  LMNB2_HUMAN     620   
3087  Q9BZL6  KPCD2_HUMAN     878   
3102  Q96PV6  LENG8_HUMAN     800   
3175  Q8IWI9   MGAP_HUMAN    3026   
3665  A6NNF4  ZN726_HUMAN     738   
3839  P49750  YLPM1_HUMAN    2146   

                                            proSequence in4084 identical_seq  \
24    MGDTFIRHIALLGFEKRFVPSQHYVYMFLVKWQDLSEKVVYRRFTE...   True         False   
261   MSHQPLSCLTEKEDSPSESTGNGPPHLAHPNLDTFTPEELLQQMKE...   True         False   
328   MHQTLCLNPESLKMSACSDFVEHIWKPGSCKNCFCLRSDHQLVAGP...   True         False   
333   MAVAVRTLQEQLEKAKESLKNVDENIRKLTGRDPNDVRPIQARLLA...   True         False   
343   MITGVFSMRLWTPVGVLTSLAYCLHQRRVALAELQEADGQCPVDRS...   True         False   
414   MAAAAPAAAAASSEAPAASATAEPEAGDQDSREVRVLQSLRGKICE...   True         False   
423   MPAMVEKGPEVSGKRRGRNNAAASASAAAASAAASAACASPAATAA...   True         False   
439   MSQTAMSETYDFLFKFLVIGNAGTGKSCLLHQFIEKKFKDDSNHTI...   True         False   
551   MSLMVSAGRGLGAVWSPTHVQVTVLQARGLRAKGPGGTSDAYAVIQ...   True         False   
563   MAALYRPGLRLNWHGLSPLGWPSCRSIQTLRVLSGDLGQLPTGIRD...   True         False   
701   MEDSGKTFSSEEEEANYWKDLAMTYKQRAENTQEELREFQEGSREY...   True         False   
748   MADDEDYEEVVEYYTEEVVYEEVPGETITKIYETTTTRTSDYEQSE...   True         False   
796   MWRCGGRQGLCVLRRLSGGHAHHRAWRWNSNRACERALQYKLGDKI...   True         False   
865   MEDEERQKKLEAGKAKLAQFRQRKAQSDGQSPSKKQKKKRKTSSSK...   True         False   
947   MAAAAASGAGGAAGAGTGGAGPAGRLLPPPAPGSPAAPAAVSPAAG...   True         False   
962   MLSRAVCGTSRQLAPVLGYLGSRQKHSLPDLPYDYGALEPHINAQI...   True         False   
1006  MANPKEKTAMCLVNELARFNRVQPQYKLLNERGPAHSKMFSVQLSL...   True         False   
1251  MAKHLKFIARTVMVQEGNVESAYRTLNRILTMDGLIEDIKHRRYYE...   True         False   
1345  MTSPAKFKKDKEIIAEYDTQVKEIRAQLTEQMKCLDQQCELRVQLL...   True         False   
1419  MDIYDTQTLGVVVFGGFMVVSAIGIFLVSTFSMKETSYEEALANQR...   True         False   
1456  MALQTLQSSWVTFRKILSHFPEELSLAFVYGSGVYRQAGPSSDQKN...   True         False   
1673  MMEAIKKKMQMLKLDKENALDRAEQAEAEQKQAEERSKQLEDELAA...   True         False   
1727  MAESLRSPRRSLYKLVGSPPWKEAFRQRCLERMRNSRDRLLNRYRQ...   True         False   
1844  MRWRTILLQYCFLLITCLLTALEAVPIDIDKTKVQNIHPVESAKIE...   True         False   
1954  MRQPYLSSREVSSSRKRWRTFPVDCVAMCGDCVEKEYPNRGNTCLE...   True         False   
1993  MKMEEAVGKVEELIESEAPPKASEQETAKEEDGSVELESQVQKDGV...   True         False   
2074  MSWGTELWDQFDSLDKHTQWGIDFLERYAKFVKERIEIEQNYAKQL...   True         False   
2223  MSLVLNDLLICCRQLEHDRATERKKEVEKFKRLIRDPETIKHLDRH...   True     

In [49]:
diff2018.to_csv("UKBseq_diff_2012_2018_ref_45diff.csv", index=False)

# [2] position of detected C & K check 

In [54]:
# FROM Pmap_make_pos_dict.py

# funx that turns list of labeled positions into sorted dictionary
# each key is index+1 of cys or lys that was labeled
# each value is chr of C or K
def get_pos_dictionary(pdcol):
    list_of_dicts = []
    for row in pdcol.tolist():
        listofkey=[]
        listofvalues=[]
        for i in row:
            v = i[0]
            listofvalues.append(v)
            k = i[1:]
            listofkey.append(k)
        di = dict(zip(listofkey, listofvalues))
        intdict = {int(oldkey): val for oldkey, val in di.items()}
        sdi = dict(sorted(intdict.items(), reverse=False))
        list_of_dicts.append(sdi)
    return list_of_dicts
def make_pos_dict():
    # import files
    cysever = "Cys2012_ever_labeled_6404.csv"
    lysever = "Lys2012_ever_labeled_9213.csv"
    cysever = pd.read_csv(cysever)
    lysever = pd.read_csv(lysever)
    # drop columns that i dont need
    cys = cysever.filter(['entry', 'pos'], axis=1)
    lys = lysever.filter(['entry', 'pos'], axis=1)
    # appending together at bottom, total shape is 
    appendedcyslys = cys.append(lys, ignore_index=True)
    print("appendedcyslys shape: ", appendedcyslys.shape)
    # creating list of positions labeled for each entry
    groupedCYSLYS = appendedcyslys.groupby('entry')['pos'].apply(list)
    groupedCYSLYS2 = pd.DataFrame(groupedCYSLYS, index=None)
    groupedCYSLYS2.reset_index(inplace=True)
    # creating column that is # of times entry has been labeled
    groupedCYSLYS2['labeled_pos_count'] = groupedCYSLYS2['pos'].str.len()
    # sorting df by decending # of times entry has been labeled
    groupedCYSLYS2_sort = groupedCYSLYS2.sort_values('labeled_pos_count',ascending=False)
    groupedCYSLYS2_sort.reset_index(inplace=True, drop=True)
    # calling funx by passing in df series (pos column)
    list_of_dicts = get_pos_dictionary(groupedCYSLYS2_sort['pos'])
    # turning list of dictionaries into a pandas column
    dictSeries = pd.Series(list_of_dicts)
    # adding dict col to original dataframe
    groupedCYSLYS2_sort['pos_dict'] = dictSeries
    # saving new file with positions of labeled cys and lys in dictionary form
    groupedCYSLYS2_sort.to_csv("dictOfPositions_allCysLys_.csv")
    return groupedCYSLYS2_sort

In [55]:
posdic12 = make_pos_dict()

appendedcyslys shape:  (15617, 2)


In [62]:
posdic12.columns = ['ID', 'pos', 'labeled_pos_count', 'pos_dict']
posdic12.head(3)

ID                                                pos  \
0  Q09666  [C108, C1833, C1900, C1967, C2162, C2806, C550...   
1  P21333  [C1018, C1157, C1165, C1260, C1353, C1453, C17...   
2  O75369  [C1087, C1158, C1280, C1326, C1617, C183, C186...   

   labeled_pos_count                                           pos_dict  
0                112  {85: 'K', 108: 'C', 128: 'K', 243: 'K', 252: '...  
1                 75  {53: 'C', 58: 'K', 87: 'K', 210: 'C', 220: 'K'...  
2                 65  {15: 'K', 16: 'K', 26: 'C', 60: 'K', 93: 'K', ...

In [61]:
checked2018.head(3)

ID   entryName  Length  \
0  Q9HAS0  NJMU_HUMAN     396   
1  Q86X76  NIT1_HUMAN     327   
2  Q9NQR4  NIT2_HUMAN     276   

                                         proSequence in4084 identical_seq  \
0  MLPSLQESMDGDEKELESSEEGGSAEERRLEPPSSSHYCLYSYRGS...   True          True   
1  MLGFITRPPHRFLSLLCPGLRIPQLSVLCAQPRPRAMAISSSSCEL...   True          True   
2  MTSFRLALIQLQISSIKSDNVTRACSFIREAATQGAKIVSLPECFN...   True          True   

   length_2012_fasta  len12_minus_len18  
0                396                  0  
1                327                  0  
2                276                  0

In [78]:
# df has ID col and protein sequences
# pos df has ID col and dictionary of positions
# merge col is shared col between two files
from ast import literal_eval # for mismap_score func
def mismap_score(df, posdf, mergecol):
    score_file = []
    # merge two df
    mer = pd.merge(df, posdf, how='inner', on=[mergecol])
    # QC merge
    print('shape ORIGINAL  df: ', df.shape)
    print('shape merge: ',mer.shape)
    print('shape ORIGINAL KEY ref: ', posdf.shape)
    print()
    for index, row in mer.iterrows():
        newrow = []
        ukb_id = row['ID']
        #ensp = row['ENSP']
        #ensp_len = row['Length']
        #ukb_len = row['UniProt_length']
        #C_abun = row['C_abundance']
        #K_abun = row['K_abundance']
        identical = row['identical_seq']
        len18 = row['Length']
        tot_tar = row['labeled_pos_count']
        pep = row['proSequence']
        posCK = row['pos_dict']


        # col order
        # newrow.append(ensp)
        newrow.append(ukb_id)
        newrow.append(identical)
        newrow.append(len18)
        #newrow.append(ukb_len)
        #newrow.append(ensp_len)
        #newrow.append(C_abun)
        #newrow.append(K_abun)
        newrow.append(tot_tar)

        # positions found counter
        count = 0
        foundC = 0
        foundK = 0
        missedC = 0
        missedK = 0

        # evaluate as dictionary
        #python_dict = literal_eval(posCK) 
        python_dict = posCK # already was dict object
        resultfound = []
        resultnotfound = []
        
        # loop thru position dict
        for key in python_dict:
            k = int(key)
            # convert pos # to index 0-based
            i = k - 1
            # if index within ensembl/or ukb protein string
            if i < int(len18):
                # get AA from peptide string
                AA = pep[i]
                # get AA from pos dic
                checker = python_dict[key]
                posid = checker + str(k)
                # match, increase count
                if AA == checker:
                    count += 1
                    resultfound.append(posid)
                    if AA == 'C':
                        foundC += 1
                    if AA == 'K':
                        foundK += 1
                if AA != checker:
                    resultnotfound.append(posid)
                    if AA == 'C':
                        missedC += 1
                    if AA == 'K':
                        missedK += 1

        # fraction correct pos
        cor_per = count/tot_tar
        miss = tot_tar - count
        # fraction missed
        miss_per = miss/tot_tar

        # add columns
        newrow.append(count) # found count
        newrow.append(miss) # missed count
        newrow.append(round(cor_per, 2))
        newrow.append(round(miss_per, 2))
        resfound = ', '.join(resultfound)
        resnotfound = ', '.join(resultnotfound)
        newrow.append(resfound)
        newrow.append(resnotfound)
        newrow.append(posCK)
        newrow.append(foundC)
        newrow.append(missedC)
        newrow.append(foundK)
        newrow.append(missedK)
        score_file.append(newrow)

    # turn list of list into dataframe
    df = pd.DataFrame(score_file)
    df.columns = ['UniProtID', 'identical_seq','len_uniprot18','total_targets','found_count','missed_count','correct_frac','missed_frac', 'positions_found', 'positions_notfound','pos_dict','found_C_count','missed_C_count','found_K_count','missed_K_count']
    #df.columns = ['UniProtID', 'identical_seq','len_uniprot18','len_ensp', 'C_abun_ukb','K_abun_ukb','total_targets','found_count','missed_count','correct_frac','missed_frac']
    return df

In [79]:
mismapdf = mismap_score(checked2018, posdic12, 'ID')

shape ORIGINAL  df:  (3976, 8)
shape merge:  (3976, 11)
shape ORIGINAL KEY ref:  (4084, 4)



In [80]:
mismapdf[mismapdf['identical_seq']== 'False']

UniProtID identical_seq  len_uniprot18  total_targets  found_count  \
24      P14598         False            390              2            2   
261     Q96CV9         False            577              2            2   
328     Q86YV5         False           1406              3            2   
333     Q9H307         False            717              2            2   
343     Q9NPH0         False            428              2            2   
414     Q15283         False            850              1            1   
423     Q9UKL0         False            485              2            0   
439     P20338         False            218              1            0   
551     Q6WKZ4         False           1283              1            1   
563     Q16822         False            640              4            4   
701     Q9NXR1         False            335              1            1   
748     P20929         False           6669              1            1   
796     Q5JRX3         False           1037              5            5   
865     Q99996         False           3907              2            0   
947     Q9Y2K2         False           1321              2            0   
962     P04179         False            222              5            5   
1006    Q9NUL3         False            570              1            1   
1251    P82921         False             87              1            1   
1345    O75044         False           1071              1            1   
1419    Q9P2E9         False           1410             13           13   
1456    Q96BW9         False            452              1            1   
1673    P06753         False            285              7            1   
1727    Q86UA6         False            219              1            1   
1844    P80303         False            420              2            2   
1954    Q8WUH1         False            139              1            1   
1993    Q96B23         False            404              1            1   
2074    Q5T0N5         False            605              1            1   
2223    Q13315         False           3056              4            4   
2234    Q66PJ3         False            421              2            0   
2395    Q99873         False            371             13            1   
2474    P13929         False            434              3            3   
2504    Q6P3S1         False            775              1            1   
2545    Q5VZ89         False           1909              2            0   
2626    Q9Y680         False            222              1            1   
2637    Q9BXW9         False           1451              1            1   
2930    Q14643         False           2758              1            1   
2948    Q8IV33         False           1275              1            1   
3001    Q13555         False            558              1            1   
3002    Q9NQT8         False           1826              1            1   
3053    Q03252         False            620              2            0   
3087    Q9BZL6         False            878              1            1   
3102    Q96PV6         False            800              1            0   
3175    Q8IWI9         False           3026              1            1   
3665    A6NNF4         False            738              1            1   
3839    P49750         False           2146              1            0   

      missed_count  correct_frac  missed_frac  \
24               0          1.00         0.00   
261              0          1.00         0.00   
328              1          0.67         0.33   
333              0          1.00         0.00   
343              0          1.00         0.00   
414              0          1.00         0.00   
423              2          0.00         1.00   
439              1          0.00         1.00   
551              0          1.00         0.00   
563              0          1.00         0.00   
701            

In [81]:
checkColumnValues(mismapdf, 'identical_seq')

  identical_seq  Count
0          True   3931
1         False     45


In [83]:
mismapdf.to_csv("UKB_2018_mismap_scored_sequences_3976.csv", index=False)

## some IDs will be dropped based on missed_fraction column
```
UniProtID	identical_seq	total_targets	found_count	missed_count	correct_frac	missed_frac	positions_found	positions_notfound
Q9UKL0	FALSE	2	0	2	0	1	0	C35, K418
P20338	FALSE	1	0	1	0	1	0	K58
Q99996	FALSE	2	0	2	0	1	0	C1966, C3868
Q9Y2K2	FALSE	2	0	2	0	1	0	K23, K31
P06753	FALSE	7	1	6	0.14	0.86	K7	K136, K149, K168, K217, K231, K248
Q66PJ3	FALSE	2	0	2	0	1	0	C159, C349
Q99873	FALSE	13	1	12	0.08	0.92	K233	C93, K98, K103, C109, K124, K135, C216, C262, K266, K303, K332, C350
Q5VZ89	FALSE	2	0	2	0	1	0	K30, C850
Q03252	FALSE	2	0	2	0	1	0	K175, K500
Q96PV6	FALSE	1	0	1	0	1	0	C529
P49750	FALSE	1	0	1	0	1	0	K1657
```

In [93]:
# IDs that had no positions found or majority of positions not found
dumpme = ["Q86YV5","Q9UKL0","P20338","Q99996","Q9Y2K2","P06753","Q66PJ3","Q99873","Q5VZ89","Q03252","Q96PV6","P49750"]

In [94]:
def addcolumnconditionalDropTrue(mapList, df, dfcol, newcol): 
    mendel = []
    for g in df[dfcol]:
        if g in mapList: 
            mendel.append("True")
        else:
            mendel.append("False")
    df.loc[:, newcol] = mendel
    df.drop(df[df[newcol] == "True"].index, inplace=True)
    df.reset_index(inplace=True, drop=True)
    print("new shape post drop: ", df.shape)
    return df

In [95]:
clean2018 = addcolumnconditionalDropTrue(dumpme, checked2018, 'ID', 'missed_majority_pos')

new shape post drop:  (3964, 9)


In [96]:
cleanmismap = addcolumnconditionalDropTrue(dumpme, mismapdf, 'UniProtID', 'missed_majority_pos')

new shape post drop:  (3964, 16)


In [97]:
checkColumnValues(clean2018, 'identical_seq')

  identical_seq  Count
0          True   3931
1         False     33


In [98]:
checkColumnValues(cleanmismap, 'identical_seq')

  identical_seq  Count
0          True   3931
1         False     33


In [100]:
cleanmismap.to_csv("UKB2018_reference_dict_mismap_3964.csv", index=False)
clean2018.to_csv("UKB2018_reference_proteome_mappedCKpositions_3964.csv", index=False)

# [3] using reference 3964 IDs with CK positions mapped to 2018, clean subset files for accurate 2018 accounting

In [102]:
idlist = list(cleanmismap.UniProtID)
print(len(idlist))
print(len(set(idlist)))

3964
3964


In [104]:
# same as above but also drops False value rows
def addcolumnconditionalDrop(mapList, df, dfcol, newcol): 
    mendel = []
    for g in df[dfcol]:
        if g in mapList: 
            mendel.append("True")
        else:
            mendel.append("False")
    df.loc[:, newcol] = mendel
    df.drop(df[df[newcol] == "False"].index, inplace = True)
    df.reset_index(inplace=True, drop=True)
    print(df.shape)
    return df

In [106]:
#df2012 = ["AllReact2012_5959.csv","ALLReact2012_grouped_entry_threshold_2807.csv","Cys2012_ever_labeled_6404.csv","CYS2012_everlabeled_gene_valueCounts_3006.csv","Cys2012_pantarget_6125.csv","Cys2012_reactive_1458.csv","CYSLYS2012_everlabeled_gene_valueCounts_4084.csv","dictOfPositions_allCysLys_4084.csv","Lys2012_ever_labeled_9213.csv","LYS2012_everlabeled_gene_valueCounts_2683.csv","Lys2012_pantarget_8257.csv"]
#df2018 = ["AllReact2018_5959.csv","ALLReact2018_grouped_entry_threshold_2807.csv","Cys2018_ever_labeled_6404.csv","CYS2018_everlabeled_gene_valueCounts_3006.csv","Cys2018_pantarget_6125.csv","Cys2018_reactive_1458.csv","CYSLYS2018_everlabeled_gene_valueCounts_4084.csv","dictOfPositions_allCysLys_4084.csv","Lys2018_ever_labeled_9213.csv","LYS2018_everlabeled_gene_valueCounts_2683.csv","Lys2018_pantarget_8257.csv"]
df2012 = ["Lys2012_reactive_4501.csv"]
df2018 = ["Lys2018_reactive_4501.csv"]
for i,d in enumerate(df2012): 
    df = pd.read_csv(d)
    print(d)
    df2 = addcolumnconditionalDrop(idlist, df, 'ID','final_refID')
    outname = df2018[i]
    df2.to_csv(outname)

Lys2012_reactive_4501.csv
(4425, 7)


## all results after filtering for 3964 IDs in 2018 with mapped positions
```bash
AllReact2012_5959.csv
(5856, 6)
ALLReact2012_grouped_entry_threshold_2807.csv
(2751, 4)
Cys2012_ever_labeled_6404.csv
(6234, 6)
CYS2012_everlabeled_gene_valueCounts_3006.csv
(2915, 3)
Cys2012_pantarget_6125.csv
(5962, 6)
Cys2012_reactive_1458.csv
(1431, 7)
CYSLYS2012_everlabeled_gene_valueCounts_4084.csv
(3964, 3)
dictOfPositions_allCysLys_4084.csv
(3964, 5)
Lys2012_ever_labeled_9213.csv
(9046, 6)
LYS2012_everlabeled_gene_valueCounts_2683.csv
(2624, 3)
Lys2012_pantarget_8257.csv
(8115, 6)
Lys2012_reactive_4501.csv
(4425, 7)
```

In [107]:
# GET ID counts in subset files

path = 'REFERENCE_2018_mappedCK/'
filelist = ["Cys2018_pantarget_5962.csv","Cys2018_reactive_1431.csv","Lys2018_pantarget_8115.csv","Lys2018_reactive_4425.csv"]

for i in filelist:
    filename = path + i
    df = pd.read_csv(filename)
    idcol = list(df.ID)
    idset = set(idcol)
    print("list lenght: ", len(idcol))
    print("set length: ", len(idset))


list lenght:  5962
set length:  2788
list lenght:  1431
set length:  984
list lenght:  8115
set length:  2370
list lenght:  4425
set length:  1545


# ACCOUNTING UPDATE with 2018 protein sequences

In [108]:
os.chdir("/Users/mariapalafox/Box Sync/CODE_DATA/dir_MAPpaper/CYS_LYS_2012_to2018UKBpositions/REFERENCE_2018_mappedCK/")

In [109]:
allreact = pd.read_csv("AllReact2018_5856.csv")
cpan = pd.read_csv("Cys2018_pantarget_5962.csv")
Cfinalreact = pd.read_csv("Cys2018_reactive_1431.csv")
kpan = pd.read_csv("Lys2018_pantarget_8115.csv")
Kfinalreact = pd.read_csv("Lys2018_reactive_4425.csv")

In [110]:
from collections import Counter
def colvalues(df, col):
    countdf = Counter(list(df[col]))
    countdic = dict(countdf)
    print(countdic)

In [111]:
colvalues(allreact, 'Threshold')
print()
colvalues(Cfinalreact, 'Threshold')
print()
colvalues(Kfinalreact, 'Threshold')
print()

{'High': 400, 'Medium': 1088, 'Low': 4368}

{'High': 97, 'Medium': 422, 'Low': 912}

{'High': 303, 'Medium': 666, 'Low': 3456}



In [112]:
checkColumnValues(cpan, "labelType")

     labelType  Count
0  panReactive   5228
1       Target    734


In [113]:
checkColumnValues(kpan,"labelType")

     labelType  Count
0  panReactive   7997
1       Target    118


In [116]:
dropUnnamed(Cfinalreact)


pos_ID  reactivity Threshold    subset      ID    pos  final_refID
0      O14733_C202        0.74      High  Cysteine  O14733   C202         True
1       P07237_C56        0.80      High  Cysteine  P07237    C56         True
2       P78417_C32        0.84      High  Cysteine  P78417    C32         True
3      Q6DCA0_C153        0.88      High  Cysteine  Q6DCA0   C153         True
4      Q9Y4X0_C175        0.88      High  Cysteine  Q9Y4X0   C175         True
5       Q00796_C45        0.89      High  Cysteine  Q00796    C45         True
6       Q9UHQ1_C99        0.89      High  Cysteine  Q9UHQ1    C99         True
7      Q9Y6I4_C302        0.90      High  Cysteine  Q9Y6I4   C302         True
8      P38606_C138        0.90      High  Cysteine  P38606   C138         True
9       Q8IY18_C91        0.92      High  Cysteine  Q8IY18    C91         True
10      P35754_C26        0.92      High  Cysteine  P35754    C26         True
11    P17927_C1387        0.94      High  Cysteine  P17927  C1387         True
12      P10599_C32        0.95      High  Cysteine  P10599    C32         True
13     Q9H7Z7_C110        0.96      High  Cysteine  Q9H7Z7   C110         True
14      P22307_C94        0.96      High  Cysteine  P22307    C94         True
15      Q9BRA2_C43        0.96      High  Cysteine  Q9BRA2    C43         True
16      P35754_C23        0.96      High  Cysteine  P35754    C23         True
17     Q6ICB0_C108        0.99      High  Cysteine  Q6ICB0   C108         True
18     Q9P0V9_C293        1.01      High  Cysteine  Q9P0V9   C293         True
19     Q9NVA2_C268        1.01      High  Cysteine  Q9NVA2   C268         True
20      Q9NS18_C77        1.01      High  Cysteine  Q9NS18    C77         True
21     O60568_C691        1.01      High  Cysteine  O60568   C691         True
22     Q92599_C271        1.01      High  Cysteine  Q92599   C271         True
23      O95881_C66        1.02      High  Cysteine  O95881    C66         True
24     P05091_C319        1.06      High  Cysteine  P05091   C319         True
25     Q9NXJ5_C149        1.06      High  Cysteine  Q9NXJ5   C149         True
26     P05091_C320        1.06      High  Cysteine  P05091   C320         True
27     P05091_C318        1.06      High  Cysteine  P05091   C318         True
28     O00267_C626        1.09      High  Cysteine  O00267   C626         True
29     P24752_C126        1.10      High  Cysteine  P24752   C126         True
...            ...         ...       ...       ...     ...    ...          ...
1401    P19838_C61       20.00       Low  Cysteine  P19838    C61         True
1402    Q96H86_C36       20.00       Low  Cysteine  Q96H86    C36         True
1403   P28838_C335       20.00       Low  Cysteine  P28838   C335         True
1404    Q96EY5_C90       20.00       Low  Cysteine  Q96EY5    C90         True
1405   Q9BQC3_C251       20.00       Low  Cysteine  Q9BQC3   C251         True
1406   Q92888_C911       20.00       Low  Cysteine  Q92888   C911         True
1407    Q8WW01_C13       20.00       Low  Cysteine  Q8WW01    C13         True
1408   P22692_C215       20.00       Low  Cysteine  P22692   C215         True
1409   Q14669_C535       20.00       Low  Cysteine  Q14669   C535         True
1410   Q05513_C161       20.00       Low  Cysteine  Q05513   C161         True
1411  Q7Z6Z7_C1832       20.00       Low  Cysteine  Q7Z6Z7  C1832         True
1412    O43707_C60       20.00       Low  Cysteine  O43707    C60         True
1413   Q9UK53_C216       20.00       Low  Cysteine  Q9UK53   C216         True
1414   Q01813_C112       20.00       Low  Cysteine  Q01813   C112         True
1415   O75369_C706       20.00       Low  Cysteine  O75369   C706         True
1416   O95347_C132       20.00       Low  Cysteine  O95347   C132         True
1417   Q69YN2_C288       20.00       Low  Cysteine  Q69YN2   C288         True
1418   Q9NVP2_C189       20.00       Low  Cysteine  Q9NVP2   C189         True
1419   Q6GMV2_C194       20.00       Low  Cysteine  Q6GMV2   C

In [117]:
dropUnnamed(Kfinalreact)

In [118]:
clow=Cfinalreact[Cfinalreact["Threshold"] == "Low"].describe()
clow.columns=['C_LOW']
cmed = Cfinalreact[Cfinalreact["Threshold"] == "Medium"].describe()
cmed.columns=['C_MED']
chigh=Cfinalreact[Cfinalreact["Threshold"] == "High"].describe()
chigh.columns=['C_HIGH']
CLM = clow.join(cmed)
CLMH = CLM.join(chigh)
print(CLMH)
CLMH.to_csv("CLMH_description_table_2018.csv")


klow = Kfinalreact[Kfinalreact["Threshold"] == "Low"].describe()
klow.columns = ['K_LOW']
kmed = Kfinalreact[Kfinalreact["Threshold"] == "Medium"].describe()
kmed.columns = ['K_MED']
khigh=Kfinalreact[Kfinalreact["Threshold"] == "High"].describe()
khigh.columns=['K_HIGH']
klm = klow.join(kmed)
klmh = klm.join(khigh)
print(klmh)
klmh.to_csv("KLMH_description_table_2018.csv")

            C_LOW       C_MED     C_HIGH
count  912.000000  422.000000  97.000000
mean     7.417007    3.851872   1.336804
std      3.056919    0.831851   0.340152
min      5.010000    2.010000   0.740000
25%      5.737500    3.192500   1.060000
50%      6.460000    4.015000   1.320000
75%      7.870000    4.572500   1.640000
max     20.000000    5.000000   1.960000
             K_LOW       K_MED      K_HIGH
count  3456.000000  666.000000  303.000000
mean      9.334259    3.518769    1.240264
std       1.341434    0.881246    0.494368
min       5.100000    2.100000    0.100000
25%       9.900000    2.700000    0.900000
50%      10.000000    3.500000    1.300000
75%      10.000000    4.300000    1.700000
max      10.000000    5.000000    2.000000


In [120]:
dropUnnamed(allreact)

In [121]:
allreact.describe()

reactivity
count  5856.000000
mean      7.427932
std       3.173233
min       0.100000
25%       5.000000
50%       8.700000
75%      10.000000
max      20.000000

In [122]:
Cfinalreact.describe()

reactivity
count  1431.000000
mean      5.953508
std       3.206224
min       0.740000
25%       4.320000
50%       5.620000
75%       6.850000
max      20.000000

In [123]:
Kfinalreact.describe()

reactivity
count  4425.000000
mean      7.904746
std       3.012153
min       0.100000
25%       5.800000
50%      10.000000
75%      10.000000
max      10.000000